mgr/dashboard: fix blockUI implementation

aaSharma14 opened the pull request:
Enable/Disable mgr modules from SMB page, Rgw multisite Page and Mgr Modules page doesn't give proper message and doesn't block the UI This PR intends to fix this issue

Fixes: https://tracker.ceph.com/issues/70557

Signed-off-by: Aashish Sharma <aasharma@redhat.com>


https://github.com/user-attachments/assets/a1ac3649-bb43-4935-8919-8881478bbd11





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

nizamial09 reviewed (approved):
thank you @aaSharma14,

aaSharma14 commented:
jenkins test dashboard

aaSharma14 commented:
jenkins test dashboard

nizamial09 commented:
looks like the PR is growing. thanks for identifying the issues @aaSharma14 . keep us updated in the PR regarding the changes that are there..

aaSharma14 commented:
@nizamial09 , This [PR](https://github.com/ceph/ceph/pull/61040) introduced a general method to enable/disable Manager Modules. There were few issues:
1. Since the method was moved from a component to a service, so using the blockUI decorator in the service was not working as expected. So we are using BlockUIService instead.
2. RGW multi-site page and CephFS Snapshot schedule page were still using the old non-general method so they were breaking.
3. Made the info alert panel for enabling the module on multi-site page similar to what we have on CephFS Snapshot Schedule page (better UI).

aaSharma14 commented:
jenkins test api

cloudbehl commented:
@aaSharma14 do we have tests around it to make sure we catch this next time it happens?

Pegonzal commented:
> 1. Since the method was moved from a component to a service, so using the blockUI decorator in the service was not working as expected. So we are using BlockUIService instead.

@aaSharma14 Could you elaborate on this? How does moving the method from component to a service stop the `blockUI` from working? You can also see the attached video in the [PR](https://github.com/ceph/ceph/pull/61040) you linked, showing the feature working without issues with `blockUI`

> 2. RGW multi-site page and CephFS Snapshot schedule page were still using the old non-general method so they were breaking. 

How does it make any sense that it broke because it was still using the non-general method? As you said [this code you are pointing to](https://github.com/ceph/ceph/blame/9ea45138fe80d0a2d6622da08b050b97834594d0/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts#L612C15-L612C15) does not use a genaral method but it's own implementation, then what does it have to do with generalized implementation itroduced in the `mgr-module.service.ts`?

You can quickly confirm that your statements are false by checking out to the commit you are pointing to `31231360fd75337305ed1f64d928312ea77b2243` and building the dashboard. It will work as expected.

I believe the issue comes from the [angular update](https://github.com/ceph/ceph/pull/59334), this is where the blocking ui and loading indicator broke

aaSharma14 commented:
> > 1. Since the method was moved from a component to a service, so using the blockUI decorator in the service was not working as expected. So we are using BlockUIService instead.
> 
> @aaSharma14 Could you elaborate on this? How does moving the method from component to a service stop the `blockUI` from working? You can also see the attached video in the [PR](https://github.com/ceph/ceph/pull/61040) you linked, showing the feature working without issues with `blockUI`
> 
> > 2. RGW multi-site page and CephFS Snapshot schedule page were still using the old non-general method so they were breaking.
> 
> How does it make any sense that it broke because it was still using the non-general method? As you said [this code you are pointing to](https://github.com/ceph/ceph/blame/9ea45138fe80d0a2d6622da08b050b97834594d0/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts#L612C15-L612C15) does not use a genaral method but it's own implementation, then what does it have to do with generalized implementation itroduced in the `mgr-module.service.ts`?
> 
> You can quickly confirm that your statements are false by checking out to the commit you are pointing to `31231360fd75337305ed1f64d928312ea77b2243` and building the dashboard. It will work as expected.
> 
> I believe the issue comes from the [angular update](https://github.com/ceph/ceph/pull/59334), this is where the blocking ui and loading indicator broke

1. @Pegonzal , I appreciate your input, but I don’t think it’s necessary to go back to previous commits and test them again. I already identified the issue and fixed it. Since the latest change was related to the commit you introduced for BlockUI, I pointed that out for clarity.

2. Regarding the RGW and cephfsSnapshot module issue, even if they had their own implementation, the functionality and code were quite similar to the method used in mgr-module.service.ts. If we are generalizing a method, I believe we should ensure all related areas are updated, not just the newly added parts.

So, I don’t think revisiting older commits is needed just to check what caused the break. The Angular update might have caused the issue—thanks for clarifying that!

Pegonzal commented:
>@Pegonzal , I appreciate your input, but I don’t think it’s necessary to go back to previous commits and test them again. 

I agree with you. But since you wrongly pointed it out first, I though it was appropiate to correct you.

> I already identified the issue and fixed it.

 It is great that you are fixing this, so thanks for that, however I don't think you have identified the issue at all, as I was asking before I didn't get the need to change from `NgBlockUI` to `BlockUIService` and the motivation for it. After researcing it I have found what I believe is the actual issue: issue https://github.com/kuuurt13/ng-block-ui/issues/275 , so looks like it's a known issue for the directives on Angular v18 for `ng-block-ui`. So this should be the reason for the changes in this PR.

>  Since the latest change was related to the commit you introduced for BlockUI, I pointed that out for clarity.

I am not sure how pointing to the wrong place would clarify anything, but feel free to elaborate on that so I can understand where should I point to next time

> Regarding the RGW and cephfsSnapshot module issue, even if they had their own implementation, the functionality and code were quite similar to the method used in mgr-module.service.ts. 

Yes, that's why I introduced a generalized approach instead of copy pasting the same code like it was being done before in multiple places

> If we are generalizing a method, I believe we should ensure all related areas are updated, not just the newly added parts.

Following the Boy Scout principle, code should be left better that initially found, I agree with you. But also, redoing things and refactoring the code is the easiest way to commit a regression, due to that and time constraints I just implemented a generalized approach that could be followed onwards and did not modified the already working pieces that were there. 


> So, I don’t think revisiting older commits is needed just to check what caused the break.

I guess that's one of the reasons why you didn't know why this broke in the first place.

Pegonzal reviewed (approved):
Just one comment from my side as I find the parameters `navigateTo`  and `navigateByUrl` overcomplicated and a bit confussing, otherwise PR LGTM, thanks @aaSharma14 for taking care of it

aaSharma14 commented:
> > 1. Since the method was moved from a component to a service, so using the blockUI decorator in the service was not working as expected. So we are using BlockUIService instead.
> 
> @aaSharma14 Could you elaborate on this? How does moving the method from component to a service stop the `blockUI` from working? You can also see the attached video in the [PR](https://github.com/ceph/ceph/pull/61040) you linked, showing the feature working without issues with `blockUI`
> 
> > 2. RGW multi-site page and CephFS Snapshot schedule page were still using the old non-general method so they were breaking.
> 
> How does it make any sense that it broke because it was still using the non-general method? As you said [this code you are pointing to](https://github.com/ceph/ceph/blame/9ea45138fe80d0a2d6622da08b050b97834594d0/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts#L612C15-L612C15) does not use a genaral method but it's own implementation, then what does it have to do with generalized implementation itroduced in the `mgr-module.service.ts`?
> 
> You can quickly confirm that your statements are false by checking out to the commit you are pointing to `31231360fd75337305ed1f64d928312ea77b2243` and building the dashboard. It will work as expected.
> 
> I believe the issue comes from the [angular update](https://github.com/ceph/ceph/pull/59334), this is where the blocking ui and loading indicator broke

1. @Pegonzal , I appreciate your input, but I don’t think it’s necessary to go back to previous commits and test them again. I already identified the issue and fixed it. Since the latest change was related to the commit you introduced for BlockUI, I pointed that out for clarity.

2. Regarding the RGW and cephfsSnapshot module issue, even if they had their own implementation, the functionality and code were quite similar to the method used in mgr-module.service.ts. If we are generalizing a method, I believe we should ensure all related areas are updated, not just the newly added parts.

So, I don’t think revisiting older commits is needed just to check what caused the break. The Angular update might have caused the issue—thanks for clarifying that!

> > @Pegonzal , I appreciate your input, but I don’t think it’s necessary to go back to previous commits and test them again.
> 
> > I agree with you. But since you wrongly pointed it out first, I though it was appropiate to correct you.
>
Thanks for the correction. 
> > I already identified the issue and fixed it.
> 
> It is great that you are fixing this, so thanks for that, however I don't think you have identified the issue at all, as I was asking before I didn't get the need to change from `NgBlockUI` to `BlockUIService` and the motivation for it. After researcing it I have found what I believe is the actual issue: issue [kuuurt13/ng-block-ui#275](https://github.com/kuuurt13/ng-block-ui/issues/275) , so looks like it's a known issue for the directives on Angular v18 for `ng-block-ui`. So this should be the reason for the changes in this PR.
> 
I completely forgot about the angular update and hence didn't consider it to be the breaking change. 
> > Since the latest change was related to the commit you introduced for BlockUI, I pointed that out for clarity.
> 
> I am not sure how pointing to the wrong place would clarify anything, but feel free to elaborate on that so I can understand where should I point to next time
> 
> > Regarding the RGW and cephfsSnapshot module issue, even if they had their own implementation, the functionality and code were quite similar to the method used in mgr-module.service.ts.
> 
> Yes, that's why I introduced a generalized approach instead of copy pasting the same code like it was being done before in multiple places
> 
> > If we are generalizing a method, I believe we should ensure all related areas are updated, not just the newly added parts.
> 
> Following the Boy Scout principle, code should be left better that initially found, I agree with you. But also, redoing things and refactoring the code is the easiest way to commit a regression, due to that and time constraints I just implemented a generalized approach that could be followed onwards and did not modified the already working pieces that were there.
> 
> > So, I don’t think revisiting older commits is needed just to check what caused the break.
> 
> I guess that's one of the reasons why you didn't know why this broke in the first place.



> > @Pegonzal , I appreciate your input, but I don’t think it’s necessary to go back to previous commits and test them again.
> 
> I agree with you. But since you wrongly pointed it out first, I though it was appropiate to correct you.
> 
Thanks for the correction.
> > I already identified the issue and fixed it.
> 
> It is great that you are fixing this, so thanks for that, however I don't think you have identified the issue at all, as I was asking before I didn't get the need to change from `NgBlockUI` to `BlockUIService` and the motivation for it. After researcing it I have found what I believe is the actual issue: issue [kuuurt13/ng-block-ui#275](https://github.com/kuuurt13/ng-block-ui/issues/275) , so looks like it's a known issue for the directives on Angular v18 for `ng-block-ui`. So this should be the reason for the changes in this PR.
> 
> > Since the latest change was related to the commit you introduced for BlockUI, I pointed that out for clarity.
> 
> I am not sure how pointing to the wrong place would clarify anything, but feel free to elaborate on that so I can understand where should I point to next time
> 
The latest code change related to BlockUI was in that PR, and at first, I thought the issue was due to how decorators work in services. However, after looking into it further, it makes sense that the actual issue is with ng-block-ui on Angular v18 (as referenced in kuuurt13/ng-block-ui#275). Given how vast Angular is, it’s not always about pointing to the wrong place. It’s part of the process to refine our understanding as we debug.
> > Regarding the RGW and cephfsSnapshot module issue, even if they had their own implementation, the functionality and code were quite similar to the method used in mgr-module.service.ts.
> 
> Yes, that's why I introduced a generalized approach instead of copy pasting the same code like it was being done before in multiple places
> 
Thanks for the efforts there. I agree that a generalized approach is better than duplicating the same logic across multiple components.
> > If we are generalizing a method, I believe we should ensure all related areas are updated, not just the newly added parts.
> 
> Following the Boy Scout principle, code should be left better that initially found, I agree with you. But also, redoing things and refactoring the code is the easiest way to commit a regression, due to that and time constraints I just implemented a generalized approach that could be followed onwards and did not modified the already working pieces that were there.
> 
I think a tracker for the rest of the cleanup would have helped just to keep the implementation consistent across the code base.
> > So, I don’t think revisiting older commits is needed just to check what caused the break.
> 
> I guess that's one of the reasons why you didn't know why this broke in the first place.
>
IMO, once the actual issue is identified, the priority should be to get the fix merged rather than testing older commits just to prove it otherwise.

aaSharma14 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/api/mgr-module.service.ts:
```diff
@@ -96,13 +97,27 @@ export class MgrModuleService {
+              );
+            }
+
+            if (!navigateTo) return;
+
+            const navigate = () => this.router.navigate([navigateTo]);
+
+            if (navigateByUrl) {
+              this.router.navigateByUrl('/', { skipLocationChange: true }).then(navigate);
+            } else {
+              navigate();
```
navigateByUrl is required on the multi-site page since on navigation to the same page, the Info message at the top about Enabling the rgw module doesn't go away automatically after the blockUI is completed.
